middlewares/mirror.go

```diff
@@ -0,0 +1,149 @@
+package middlewares
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	"github.com/vulcand/oxy/forward"
+	"github.com/vulcand/oxy/utils"
+	"net/url"
+)
+
+//Mirror is a middleware that provides the request mirroring capabilities
+type Mirror struct {
+	BackendURL     *url.URL
+	RequestHeaders map[string]string
+	SampleRate     uint64
+	mirror         *forward.Forwarder
+	counter        uint64
+}
+
+//NewMirrorMiddleware initializes the Mirror for the request mirroring
+func NewMirrorMiddleware(frontend *types.Frontend, backend *types.Backend) (*Mirror, error) {
+	mirror := frontend.Mirror
+	// Default sample rate is every request.
+	var sampleRate uint64 = 1
+	if mirror.SampleRate > 0 {
+		sampleRate = uint64(mirror.SampleRate)
+	}
+	mirrorURL, err := url.Parse(backend.Servers["mirror"].URL)
+	if err != nil {
+		return nil, err
+	}
+
+	forwarder, err := forward.New(
+	//forward.Stream(true),
+	//forward.PassHostHeader(frontend.PassHostHeader),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &Mirror{BackendURL: mirrorURL,
+			RequestHeaders: mirror.RequestHeaders,
+			SampleRate:     sampleRate,
+			mirror:         forwarder},
+		nil
+}
+
+func (m *Mirror) ServeHTTP(w http.ResponseWriter, req *http.Request, next http.HandlerFunc) {
+	if m.counter%m.SampleRate == 0 {
+		atomic.AddUint64(&m.counter, 1)
+		// Copy the request and forward to 'next' and to 'mirror'
+		mReq, err := m.copyRequest(req)
+		if err != nil {
+			log.Errorf("Encountered an error while cloning the request: %v", err)
+		}
+
+		// We don't care about the response from the mirror, so we can just launch it in a safe go routine
+		// In case of failure, will log the error
+		safe.GoWithRecover(func() {
+			mRw := createDummyResponseWriter()
+			m.mirror.ServeHTTP(mRw, mReq)
+		}, func(err interface{}) {
+			log.Errorf("Error in request mirroring routine: %s", err)
+		})
+
+		next.ServeHTTP(w, req)
+	} else {
+		// no mirroring
+		next.ServeHTTP(w, req)
+	}
+}
+
+func (m *Mirror) copyRequest(req *http.Request) (*http.Request, error) {
+	var body []byte
+	o := *req
+
+	if req.Body != nil {
+		b, err := ioutil.ReadAll(req.Body)
```

Comment:
I can see a couple of problems with consuming the request body here:
- won't this "drain" the request body and therefore make it appear as empty or at the end for other consumers of the request body like the original request forwarding?
- this is still in the synchronous path of the original request execution and therefore it would block until a request is forwarded until the whole request body is consumed. This can delay proxying significantly for slow clients, like mobile phones with low bandwidth.

Reply:
This shouldn't drain the request body as the `Body` is being read into a `body` byte array, which later gets assigned to the `Body` of the original request and the clone individually.  So by the time `next.ServeHTTP` is being called, each request has a proper body.

The choice of doing request mirroring assumes that the developers are willing to "pay" for the capability with the latency introduced by the need to buffer up (read in) the request data in order to be able to fork it to both destinations.  Do you have any proposed techniques that might improve this?